packages/serverless-workflow-editor/src/monaco/augmentation/MonacoEditorApi.ts

```diff
@@ -23,28 +23,53 @@ export interface MonacoEditorApi {
 
   undo: () => void;
   redo: () => void;
+
+  getContent: () => string;
+  pushUndoStop: () => void;
+}
+
+export enum MonacoEditorOperation {
+  UNDO,
+  REDO,
+  EDIT,
 }
 
 export class DefaultMonacoEditor implements MonacoEditorApi {
   private readonly model: editor.ITextModel;
 
   private editor: editor.IStandaloneCodeEditor;
   private readonly augmentation: MonacoAugmentation;
+  private readonly onContentChange: (content: string, operation: MonacoEditorOperation) => void;
 
-  constructor(content: string, onContentChange: (content: string) => void, augmentation: MonacoAugmentation) {
+  constructor(
+    content: string,
+    onContentChange: (content: string, operation: MonacoEditorOperation) => void,
+    augmentation: MonacoAugmentation
+  ) {
     this.model = editor.createModel(augmentation.language.getDefaultContent(content), augmentation.language.languageId);
-    this.model.onDidChangeContent((event) => onContentChange(this.model.getValue()));
+    this.model.onDidChangeContent((event) => {
+      if (!event.isUndoing && !event.isRedoing) {
+        onContentChange(this.model.getValue(), MonacoEditorOperation.EDIT);
+      }
+    });
+    this.onContentChange = onContentChange;
     this.augmentation = augmentation;
   }
 
   redo(): void {
+    this.editor?.focus();
     this.editor?.trigger("whatever...", "redo", null);
```

Comment:
```suggestion
    this.editor?.trigger("??????", "redo", null);
```

What's the first parameter here? I think we could use something better than `whatever` 😝